zio-http/src/main/scala/zio/http/Request.scala

```diff
@@ -107,8 +107,41 @@ final case class Request(
   def unnest(prefix: Path): Request =
     copy(url = self.url.copy(path = self.url.path.unnest(prefix)))
 
+  /**
+   * Returns the cookie with the given name if it exists.
+   */
   def cookie(name: String): Option[Cookie] =
-    header(Header.Cookie).map(_.value).flatMap(_.filter(_.name == name).headOption)
+    cookies.find(_.name == name)
+
+  /**
+   * Uses the cookie with the given name if it exists and runs `f` afterwards.
+   */
+  def cookieWithZIO[R, A](name: String)(f: Cookie => ZIO[R, Throwable, A])(implicit
+    trace: Trace,
+  ): ZIO[R, Throwable, A] =
+    cookieWithOrFailImpl(name)(identity)(f)
+
+  /**
+   * Uses the cookie with the given name if it exists and runs `f` afterwards.
+   *
+   * Also, you can replace a `NoSuchElementException` from an absent cookie with
+   * `E`.
+   */
+  def cookieWithOrFail[R, E, A](name: String)(missingCookieError: E)(f: Cookie => ZIO[R, E, A])(implicit
+    trace: Trace,
+  ): ZIO[R, E, A] =
+    cookieWithOrFailImpl(name)(_ => missingCookieError)(f)
+
+  private def cookieWithOrFailImpl[R, E, A](name: String)(e: Throwable => E)(f: Cookie => ZIO[R, E, A])(implicit
+    trace: Trace,
+  ): ZIO[R, E, A] =
+    ZIO.getOrFailWith(e(new java.util.NoSuchElementException(s"cookie doesn't exist: $name")))(cookie(name)).flatMap(f)
```

Comment:
This will create an exception with stacktrace just to be thrown away immediately. This does not seem to be efficient.
I'd change it just simply to 
```suggestion
    cookie(name) match {
      case Some(cookie) => f(e)
      case None => ZIO.fail(e)
    }
```
where e is not a function

Reply:
Added your suggestions, thank you!